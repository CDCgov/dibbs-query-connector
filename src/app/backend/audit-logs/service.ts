"use server";

import dbService from "../db/client";

export type LogEntry = {
  actionType: string;
  auditChecksum: string;
  author: string;
  auditMessage: { [argLabel: string]: string };
  createdAt: Date;
};

class AuditLogService {
  static async getAuditLogs() {
    const auditQuery = "SELECT * FROM audit_logs ORDER BY created_at DESC;";
    const results = await dbService.query(auditQuery);
    return results.rows;
  }
}

export const getAuditLogs = AuditLogService.getAuditLogs;
