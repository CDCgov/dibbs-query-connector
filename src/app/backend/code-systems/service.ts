"use server";
import { Bundle, OperationOutcome, Parameters, ValueSet } from "fhir/r4";
import {
  ersdToDibbsConceptMap,
  MISSING_API_KEY_LITERAL,
} from "../../constants";
import { encode } from "base-64";
import { indexErsdByOid } from "../db-creation/lib";
import {} from "../db-creation/service";

export type ErsdOrVsacResponse = Bundle | Parameters | OperationOutcome;

type ersdCondition = {
  code: string;
  system: string;
  text: string;
  valueset_id: string;
};

export type OidData = {
  oids: Array<string>;
  oidToErsdType: Map<string, string>;
  conditions: Array<ersdCondition>;
};

/**
 * Performs eRSD querying and parsing to generate OIDs and extract clinical
 * concepts from the eRSD. First, a call is made to the eRSD, which is filtered
 * for valuesets. These valusets are used to create a mapping between OID and
 * clinical service code, as well as to actually compile the list of OIDs of all
 * valuesets in the pull-down. Then, these two entities are bundled together
 * as a data structure for future scripting.
 * @returns An OidData object containing the OIDs extracted as well as the
 * clinical service type associated with each.
 */
export async function getOidsFromErsd() {
  console.log("Fetching and parsing eRSD.");
  const ersd = await getERSD();
  const valuesets = (ersd as unknown as Bundle)["entry"]?.filter(
    (e) => e.resource?.resourceType === "ValueSet",
  );

  const { nonUmbrellaValueSets, oidToErsdType } = indexErsdByOid(valuesets);
  // Build up a mapping of OIDs to eRSD clinical types

  let conditionExtractor: Array<ersdCondition> = [];
  nonUmbrellaValueSets.reduce((acc: Array<ersdCondition>, vs: ValueSet) => {
    const conditionSchemes = vs.useContext?.filter(
      (context) =>
        !(context.valueCodeableConcept?.coding || [])[0].system?.includes(
          "us-ph-usage-context",
        ),
    );
    (conditionSchemes || []).forEach((usc) => {
      const ersdCond: ersdCondition = {
        code: (usc.valueCodeableConcept?.coding || [])[0].code || "",
        system: (usc.valueCodeableConcept?.coding || [])[0].system || "",
        text: usc.valueCodeableConcept?.text || "",
        valueset_id: vs.id || "",
      };
      conditionExtractor.push(ersdCond);
    });
    return conditionExtractor;
  }, conditionExtractor);

  // Make sure to take out the umbrella value sets from the ones we try to insert
  let oids = valuesets?.map((vs) => vs.resource?.id);
  oids = oids?.filter(
    (oid) => !Object.keys(ersdToDibbsConceptMap).includes(oid || ""),
  );
  return {
    oids: oids,
    oidToErsdType: oidToErsdType,
    conditions: conditionExtractor,
  } as OidData;
}

/**
 * Fetches the eRSD Specification from the eRSD API. This function requires an API key
 * to access the eRSD API. The API key can be obtained at https://ersd.aimsplatform.org/#/api-keys.
 * @param eRSDVersion - The version of the eRSD specification to retrieve. Defaults to v2.
 * @returns The eRSD Specification as a FHIR Bundle or an OperationOutcome if an error occurs.
 */
export async function getERSD(
  eRSDVersion: number = 3,
): Promise<ErsdOrVsacResponse> {
  const ERSD_API_KEY = process.env.ERSD_API_KEY;
  if (!ERSD_API_KEY) {
    throw Error(
      `ERSD API Key not set. Please refer to the documentation below to get your ERSD key to continue`,
      {
        cause: MISSING_API_KEY_LITERAL,
      },
    );
  }

  const eRSDUrl = `https://ersd.aimsplatform.org/api/ersd/v${eRSDVersion}specification?format=json&api-key=${ERSD_API_KEY}`;
  const response = await fetch(eRSDUrl);

  if (response.status === 200) {
    const data = (await response.json()) as Bundle;

    return data;
  } else {
    return {
      resourceType: "OperationOutcome",
      issue: [
        {
          severity: "error",
          code: "processing",
          diagnostics: `Failed to retrieve data from eRSD: ${response.status} ${response.statusText}`,
        },
      ],
    } as OperationOutcome;
  }
}

/**
 * Fetches the VSAC Value Sets and supporting code systems information for a given OID
 * as a FHIR bundle. This function requires a UMLS API Key which must be obtained as a
 * Metathesaurus License. See https://www.nlm.nih.gov/vsac/support/usingvsac/vsacfhirapi.html
 * for authentication instructions.
 * @param oid The OID whose value sets to retrieve.
 * @param searchStructType Optionally, a flag to identify what type of
 * search to perform, since VSAC can be used for value sets as well as conditions.
 * @param codeSystem Optional parameter for use with condition querying.
 * @returns The value sets as a FHIR bundle, or an Operation Outcome if there is an error.
 */
export async function getVSACValueSet(
  oid: string,
  searchStructType: string = "valueset",
  codeSystem?: string,
): Promise<ErsdOrVsacResponse> {
  const username: string = "apikey";
  const umlsKey = process.env.UMLS_API_KEY;

  if (!umlsKey) {
    throw Error(
      "UMLS API Key not set. Please refer to the documentation below on how to get your UMLS API key before continuing",
      {
        cause: MISSING_API_KEY_LITERAL,
      },
    );
  }

  const vsacUrl: string =
    searchStructType === "valueset"
      ? `https://cts.nlm.nih.gov/fhir/ValueSet/${oid}`
      : `https://cts.nlm.nih.gov/fhir/CodeSystem/$lookup?system=${codeSystem}&code=${oid}`;
  const response = await fetch(vsacUrl, {
    method: "get",
    headers: new Headers({
      Authorization: "Basic " + encode(username + ":" + umlsKey),
      "Content-Type": "application/fhir+json",
    }),
  });
  if (response.status === 200) {
    const data = (await response.json()) as Bundle;
    return data;
  } else {
    const diagnosticIssue = await response.json().then((r) => {
      try {
        return r?.issue[0]?.diagnostics;
      } catch {
        return r;
      }
    });
    return {
      resourceType: "OperationOutcome",
      issue: [
        {
          severity: "error",
          code: "processing",
          diagnostics: `${response.status}: ${diagnosticIssue}`,
        },
      ],
    } as OperationOutcome;
  }
}

/**
 * helper function to generate VSAC promises
 *
 * @param oidsToFetch - OIDs from the eRSD to query from VSAC
 * @returns Promises to resolve that will give you the requested VSAC valuesets
 */
export async function generateBatchVsacPromises(oidsToFetch: string[]) {
  const valueSetPromises = Promise.allSettled(
    oidsToFetch.map(async (oid) => {
      try {
        // First, we'll pull the value set from VSAC and map it to our representation
        const vs = await getVSACValueSet(oid);
        return { vs: vs, oid: oid };
      } catch (e) {
        let message = `Fetch for VSAC value set with oid ${oid} failed`;
        if (e instanceof Error) {
          message = message + ` with error message: ${e.message}`;
        }
        return Promise.reject(new Error(message));
      }
    }),
  );

  return valueSetPromises;
}
