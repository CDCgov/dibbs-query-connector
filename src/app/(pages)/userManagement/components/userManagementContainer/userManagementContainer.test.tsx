import { waitFor, screen, render } from "@testing-library/react";
import * as UserManagementBackend from "@/app/backend/user-management";
import * as UserGroupManagementBackend from "@/app/backend/usergroup-management";
import * as QueryBuildingBackend from "@/app/backend/query-building/service";
import { UserRole } from "@/app/models/entities/users";
import { renderWithUser, RootProviderMock } from "@/app/tests/unit/setup";
import UsersTable from "./userManagementContainer";
import {
  mockSuperAdmin,
  mockAdmin,
  mockGroupBasic,
  mockPermissionsTab,
  mockQueryWithGroups,
} from "../../test-utils";

jest.mock("next-auth/react");

jest.mock("@/app/backend/user-management", () => ({
  getAllUsers: jest.fn().mockResolvedValue({ items: [], totalItems: 0 }),
  getUserRole: jest.fn(),
}));

jest.mock("@/app/backend/query-building/service", () => ({
  getCustomQueries: jest.fn().mockResolvedValue([]),
}));

jest.mock("@/app/backend/usergroup-management", () => ({
  getAllUserGroups: jest.fn().mockResolvedValue({ items: [], totalItems: 0 }),
  getSingleQueryGroupAssignments: jest
    .fn()
    .mockResolvedValue({ items: [], totalItems: 0 }),
}));

jest.mock("@/app/backend/query-building/service", () => ({
  getCustomQueries: jest.fn().mockResolvedValue([]),
}));

jest.mock(
  "@/app/ui/components/withAuth/WithAuth",
  () =>
    ({ children }: React.PropsWithChildren) => <div>{children}</div>,
);

jest.mock(
  "@/app/(pages)/userManagement/components/userModal/userModal",
  () => () => <div>Modal Mock</div>,
);
describe("Super Admin view of Users Table", () => {
  const role = UserRole.SUPER_ADMIN;
  it("renders error message when no users are found", async () => {
    render(
      <RootProviderMock currentPage="/userManagement">
        <UsersTable role={role} />
      </RootProviderMock>,
    );

    await waitFor(() => {
      expect(screen.queryByText("Loading")).not.toBeInTheDocument();
    });

    expect(screen.getByText("No users found")).toBeInTheDocument();
    expect(document.body).toMatchSnapshot();
  });

  it("renders table view after content is loaded", async () => {
    jest.spyOn(UserManagementBackend, "getAllUsers").mockResolvedValueOnce({
      items: [mockAdmin, mockSuperAdmin],
      totalItems: 2,
    });

    render(
      <RootProviderMock currentPage="/userManagement">
        <UsersTable role={role} />
      </RootProviderMock>,
    );
    await waitFor(() => {
      expect(screen.queryByText("Loading")).not.toBeInTheDocument();
    });

    expect(document.body).toMatchSnapshot();

    if (mockPermissionsTab.renderContent) {
      render(mockPermissionsTab.renderContent());
    }

    expect(document.body).toHaveTextContent("Potter, Harry");
    expect(document.body).toHaveTextContent("Order of the Phoenix");
    expect(document.body).toMatchSnapshot();

    jest.restoreAllMocks();
  });

  it("renders content on tab click", async () => {
    jest.spyOn(QueryBuildingBackend, "getCustomQueries").mockResolvedValue([]);

    jest.spyOn(UserManagementBackend, "getAllUsers").mockResolvedValueOnce({
      items: [mockAdmin, mockSuperAdmin],
      totalItems: 2,
    });
    jest
      .spyOn(UserGroupManagementBackend, "getAllUserGroups")
      .mockResolvedValueOnce({
        items: [mockGroupBasic],
        totalItems: 1,
      });
    jest
      .spyOn(QueryBuildingBackend, "getCustomQueries")
      .mockResolvedValueOnce([mockQueryWithGroups]);
    const { user } = renderWithUser(
      <RootProviderMock currentPage="/userManagement">
        <UsersTable role={role} />
      </RootProviderMock>,
    );
    await waitFor(() => {
      expect(screen.queryByText("Loading")).not.toBeInTheDocument();
    });

    expect(document.body).toMatchSnapshot();
    const usersTab = document.getElementsByClassName("tab__active")[0];
    expect(usersTab).toHaveTextContent("Users");
    expect(document.body).toHaveTextContent("Permissions");

    const groupsTab = document.getElementsByClassName("tab")[0];
    expect(groupsTab).toHaveTextContent("User groups");

    await user.click(groupsTab);
    expect(groupsTab).toHaveClass("tab__active");

    expect(document.body).toHaveTextContent("Create group");
    expect(document.body).toMatchSnapshot();

    jest.restoreAllMocks();
  });
});

describe("Admin view of Users Table", () => {
  const role = UserRole.ADMIN;
  it("renders error message when no user groups are found", async () => {
    render(
      <RootProviderMock currentPage="/userManagement">
        <UsersTable role={role} />
      </RootProviderMock>,
    );

    await waitFor(() => {
      expect(screen.queryByText("Loading")).not.toBeInTheDocument();
    });

    expect(screen.getByText("No user groups found")).toBeInTheDocument();
    expect(document.body).toMatchSnapshot();
    jest.restoreAllMocks();
  });
});
