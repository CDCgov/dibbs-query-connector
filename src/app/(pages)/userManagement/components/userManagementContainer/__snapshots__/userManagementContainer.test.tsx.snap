// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Admin view of Users Table renders error message when no user groups are found 1`] = `
<body>
  <div>
    <div
      class="main-container__wide"
    >
      <div
        class="empty-response"
      >
        <div
          class="empty-response display-flex flex-column flex-align-start"
        >
          No user groups found
          <button
            class="usa-button styles.createQueryButton margin-bottom-3"
            data-testid="button"
            style="margin-left: 1px; margin-top: 1.5rem;"
            type="button"
          >
            Create group
          </button>
        </div>
      </div>
      <div>
        Modal Mock
      </div>
      <div>
        <div
          aria-label="drawer-container"
          class="drawer closed width35"
          data-testid="drawer-open-false"
          id="drawer-container"
          role="dialog"
        >
          <div
            class="display-none"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Super Admin view of Users Table renders content on tab click 1`] = `
<body>
  <div>
    <div
      class="main-container__wide"
    >
      <div
        class="border-base border-bottom-1px margin-bottom-3"
      >
        <button
          class="usa-button--unstyled tab__active"
        >
          Users
        </button>
        <button
          class="usa-button--unstyled tab"
        >
          User groups
        </button>
      </div>
      <button
        class="usa-button createQueryButton margin-bottom-3"
        data-testid="button"
        type="button"
      >
        Add user
      </button>
      <div
        class="userPermissionsTable customizeTableWrapper"
      >
        <table
          class="usa-table usa-table--borderless customizeTable"
          data-testid="table"
        >
          <thead>
            <tr>
              <th
                class="grid-col-5"
              >
                Name
              </th>
              <th
                class="grid-col-3"
              >
                Permissions
              </th>
              <th
                class="grid-col-4"
              >
                User groups
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              id="456"
              tabindex="0"
            >
              <td>
                <div
                  class="buttonHoverGroup"
                >
                  <span>
                    Potter, Lily 
                  </span>
                  <button
                    class="usa-button usa-button--unstyled hoverBtn unstyled-button-container usa-button--unstyled text-bold text-no-underline"
                    data-testid="button"
                    type="button"
                  >
                    <span
                      class="icon-text padding-right-4 display-flex flex-align-center"
                    >
                      <svg
                        aria-label="Pencil icon indicating edit ability"
                        class="usa-icon height-3 width-3"
                        focusable="false"
                        height="1em"
                        role="img"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 0 0 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                        />
                      </svg>
                      <span
                        class="padding-left-05"
                        data-testid="edit-group-456"
                      >
                        Rename
                      </span>
                    </span>
                  </button>
                </div>
              </td>
              <td
                class="roleDropdown"
              >
                <label
                  class="usa-label usa-sr-only"
                  data-testid="label"
                  for="role-select-456"
                >
                  User role
                </label>
                <select
                  class="usa-select"
                  data-testid="role-select-456"
                  id="role-select-456"
                >
                  <option
                    value="Super Admin"
                  >
                    Super Admin
                  </option>
                  <option
                    selected=""
                    value="Admin"
                  >
                    Admin
                  </option>
                  <option
                    value="Standard"
                  >
                    Standard
                  </option>
                </select>
              </td>
              <td>
                <button
                  aria-description="Edit Order of the Phoenix members"
                  class="usa-button usa-button--unstyled margin-right-2 text-no-underline"
                  data-testid="button"
                  type="button"
                >
                  <span>
                    Order of the Phoenix
                  </span>
                </button>
              </td>
            </tr>
            <tr
              id="123"
              tabindex="0"
            >
              <td>
                <div
                  class="buttonHoverGroup"
                >
                  <span>
                    Potter, Harry 
                  </span>
                  <button
                    class="usa-button usa-button--unstyled hoverBtn unstyled-button-container usa-button--unstyled text-bold text-no-underline"
                    data-testid="button"
                    type="button"
                  >
                    <span
                      class="icon-text padding-right-4 display-flex flex-align-center"
                    >
                      <svg
                        aria-label="Pencil icon indicating edit ability"
                        class="usa-icon height-3 width-3"
                        focusable="false"
                        height="1em"
                        role="img"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 0 0 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                        />
                      </svg>
                      <span
                        class="padding-left-05"
                        data-testid="edit-group-123"
                      >
                        Rename
                      </span>
                    </span>
                  </button>
                </div>
              </td>
              <td
                class="roleDropdown"
              >
                <label
                  class="usa-label usa-sr-only"
                  data-testid="label"
                  for="role-select-123"
                >
                  User role
                </label>
                <select
                  class="usa-select"
                  data-testid="role-select-123"
                  id="role-select-123"
                >
                  <option
                    selected=""
                    value="Super Admin"
                  >
                    Super Admin
                  </option>
                  <option
                    value="Admin"
                  >
                    Admin
                  </option>
                  <option
                    value="Standard"
                  >
                    Standard
                  </option>
                </select>
              </td>
              <td>
                --
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div>
        Modal Mock
      </div>
      <div>
        <div
          aria-label="drawer-container"
          class="drawer closed width35"
          data-testid="drawer-open-false"
          id="drawer-container"
          role="dialog"
        >
          <div
            class="display-none"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Super Admin view of Users Table renders content on tab click 2`] = `
<body>
  <div>
    <div
      class="main-container__wide"
    >
      <div
        class="border-base border-bottom-1px margin-bottom-3"
      >
        <button
          class="usa-button--unstyled tab"
        >
          Users
        </button>
        <button
          class="usa-button--unstyled tab__active"
        >
          User groups
        </button>
      </div>
      <button
        class="usa-button createQueryButton margin-bottom-3"
        data-testid="button"
        type="button"
      >
        Create group
      </button>
      <div
        class="userGroupsTable customizeTableWrapper"
      >
        <table
          class="usa-table usa-table--borderless customizeTable"
          data-testid="table"
        >
          <thead>
            <tr>
              <th
                class="grid-col-5"
              >
                Name
              </th>
              <th
                class="grid-col-3"
              >
                Members
              </th>
              <th
                class="grid-col-4"
              >
                Assigned queries
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              id="123"
              tabindex="0"
            >
              <td>
                <div
                  class="buttonHoverGroup"
                >
                  <span>
                    Hogwarts
                  </span>
                  <button
                    class="usa-button usa-button--unstyled hoverBtn unstyled-button-container usa-button--unstyled text-bold text-no-underline"
                    data-testid="button"
                    type="button"
                  >
                    <span
                      class="icon-text padding-right-4 display-flex flex-align-center"
                    >
                      <svg
                        aria-label="Pencil icon indicating edit ability"
                        class="usa-icon height-3 width-3"
                        focusable="false"
                        height="1em"
                        role="img"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 0 0 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                        />
                      </svg>
                      <span
                        class="padding-left-05"
                        data-testid="edit-group-123"
                      >
                        Rename
                      </span>
                    </span>
                  </button>
                </div>
              </td>
              <td>
                <button
                  aria-description="Edit Hogwarts members"
                  class="usa-button usa-button--unstyled drawerButton margin-right-2 text-no-underline"
                  data-testid="edit-member-list-0"
                  type="button"
                >
                  1 member
                </button>
              </td>
              <td
                class="grid-col-2"
              >
                <div
                  class="buttonHoverGroup"
                >
                  <button
                    aria-description="Edit Hogwarts queries"
                    class="usa-button usa-button--unstyled drawerButton text-no-underline"
                    data-testid="edit-query-list-0"
                    type="button"
                  >
                    0 queries
                  </button>
                  <button
                    class="usa-button usa-button--unstyled hoverBtn unstyled-button-container usa-button--unstyled text-bold text-no-underline destructive-primary"
                    data-testid="button"
                    type="button"
                  >
                    <span
                      class="icon-text padding-right-4 display-flex flex-align-center"
                    >
                      <svg
                        aria-label="Trash icon indicating deletion of disease"
                        class="usa-icon height-3 width-3"
                        focusable="false"
                        height="1em"
                        role="img"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                        />
                      </svg>
                      <span
                        class="padding-left-05"
                      >
                        Delete
                      </span>
                    </span>
                  </button>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div>
        Modal Mock
      </div>
      <div>
        <div
          aria-label="drawer-container"
          class="drawer closed width35"
          data-testid="drawer-open-false"
          id="drawer-container"
          role="dialog"
        >
          <div
            class="display-none"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Super Admin view of Users Table renders error message when no users are found 1`] = `
<body>
  <div>
    <div
      class="main-container__wide"
    >
      <div
        class="border-base border-bottom-1px margin-bottom-3"
      >
        <button
          class="usa-button--unstyled tab__active"
        >
          Users
        </button>
        <button
          class="usa-button--unstyled tab"
        >
          User groups
        </button>
      </div>
      <button
        class="usa-button createQueryButton margin-bottom-3"
        data-testid="button"
        type="button"
      >
        Add user
      </button>
      <div
        class="empty-response"
      >
        No users found
      </div>
      <div>
        Modal Mock
      </div>
      <div>
        <div
          aria-label="drawer-container"
          class="drawer closed width35"
          data-testid="drawer-open-false"
          id="drawer-container"
          role="dialog"
        >
          <div
            class="display-none"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Super Admin view of Users Table renders table view after content is loaded 1`] = `
<body>
  <div>
    <div
      class="main-container__wide"
    >
      <div
        class="border-base border-bottom-1px margin-bottom-3"
      >
        <button
          class="usa-button--unstyled tab__active"
        >
          Users
        </button>
        <button
          class="usa-button--unstyled tab"
        >
          User groups
        </button>
      </div>
      <button
        class="usa-button createQueryButton margin-bottom-3"
        data-testid="button"
        type="button"
      >
        Add user
      </button>
      <div
        class="empty-response"
      >
        No users found
      </div>
      <div>
        Modal Mock
      </div>
      <div>
        <div
          aria-label="drawer-container"
          class="drawer closed width35"
          data-testid="drawer-open-false"
          id="drawer-container"
          role="dialog"
        >
          <div
            class="display-none"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`Super Admin view of Users Table renders table view after content is loaded 2`] = `
<body>
  <div>
    <div
      class="main-container__wide"
    >
      <div
        class="border-base border-bottom-1px margin-bottom-3"
      >
        <button
          class="usa-button--unstyled tab__active"
        >
          Users
        </button>
        <button
          class="usa-button--unstyled tab"
        >
          User groups
        </button>
      </div>
      <button
        class="usa-button createQueryButton margin-bottom-3"
        data-testid="button"
        type="button"
      >
        Add user
      </button>
      <div
        class="userPermissionsTable customizeTableWrapper"
      >
        <table
          class="usa-table usa-table--borderless customizeTable"
          data-testid="table"
        >
          <thead>
            <tr>
              <th
                class="grid-col-5"
              >
                Name
              </th>
              <th
                class="grid-col-3"
              >
                Permissions
              </th>
              <th
                class="grid-col-4"
              >
                User groups
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              id="456"
              tabindex="0"
            >
              <td>
                <div
                  class="buttonHoverGroup"
                >
                  <span>
                    Potter, Lily 
                  </span>
                  <button
                    class="usa-button usa-button--unstyled hoverBtn unstyled-button-container usa-button--unstyled text-bold text-no-underline"
                    data-testid="button"
                    type="button"
                  >
                    <span
                      class="icon-text padding-right-4 display-flex flex-align-center"
                    >
                      <svg
                        aria-label="Pencil icon indicating edit ability"
                        class="usa-icon height-3 width-3"
                        focusable="false"
                        height="1em"
                        role="img"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 0 0 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                        />
                      </svg>
                      <span
                        class="padding-left-05"
                        data-testid="edit-group-456"
                      >
                        Rename
                      </span>
                    </span>
                  </button>
                </div>
              </td>
              <td
                class="roleDropdown"
              >
                <label
                  class="usa-label usa-sr-only"
                  data-testid="label"
                  for="role-select-456"
                >
                  User role
                </label>
                <select
                  class="usa-select"
                  data-testid="role-select-456"
                  id="role-select-456"
                >
                  <option
                    value="Super Admin"
                  >
                    Super Admin
                  </option>
                  <option
                    selected=""
                    value="Admin"
                  >
                    Admin
                  </option>
                  <option
                    value="Standard"
                  >
                    Standard
                  </option>
                </select>
              </td>
              <td>
                <button
                  aria-description="Edit Order of the Phoenix members"
                  class="usa-button usa-button--unstyled margin-right-2 text-no-underline"
                  data-testid="button"
                  type="button"
                >
                  <span>
                    Order of the Phoenix
                  </span>
                </button>
              </td>
            </tr>
            <tr
              id="123"
              tabindex="0"
            >
              <td>
                <div
                  class="buttonHoverGroup"
                >
                  <span>
                    Potter, Harry 
                  </span>
                  <button
                    class="usa-button usa-button--unstyled hoverBtn unstyled-button-container usa-button--unstyled text-bold text-no-underline"
                    data-testid="button"
                    type="button"
                  >
                    <span
                      class="icon-text padding-right-4 display-flex flex-align-center"
                    >
                      <svg
                        aria-label="Pencil icon indicating edit ability"
                        class="usa-icon height-3 width-3"
                        focusable="false"
                        height="1em"
                        role="img"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 0 0 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                        />
                      </svg>
                      <span
                        class="padding-left-05"
                        data-testid="edit-group-123"
                      >
                        Rename
                      </span>
                    </span>
                  </button>
                </div>
              </td>
              <td
                class="roleDropdown"
              >
                <label
                  class="usa-label usa-sr-only"
                  data-testid="label"
                  for="role-select-123"
                >
                  User role
                </label>
                <select
                  class="usa-select"
                  data-testid="role-select-123"
                  id="role-select-123"
                >
                  <option
                    selected=""
                    value="Super Admin"
                  >
                    Super Admin
                  </option>
                  <option
                    value="Admin"
                  >
                    Admin
                  </option>
                  <option
                    value="Standard"
                  >
                    Standard
                  </option>
                </select>
              </td>
              <td>
                --
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div>
        Modal Mock
      </div>
      <div>
        <div
          aria-label="drawer-container"
          class="drawer closed width35"
          data-testid="drawer-open-false"
          id="drawer-container"
          role="dialog"
        >
          <div
            class="display-none"
          />
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      class="userPermissionsTable customizeTableWrapper"
    >
      <table
        class="usa-table usa-table--borderless customizeTable"
        data-testid="table"
      >
        <thead>
          <tr>
            <th
              class="grid-col-5"
            >
              Name
            </th>
            <th
              class="grid-col-3"
            >
              Permissions
            </th>
            <th
              class="grid-col-4"
            >
              User groups
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            id="456"
            tabindex="0"
          >
            <td>
              <div
                class="buttonHoverGroup"
              >
                <span>
                  Potter, Lily 
                </span>
                <button
                  class="usa-button usa-button--unstyled hoverBtn unstyled-button-container usa-button--unstyled text-bold text-no-underline"
                  data-testid="button"
                  type="button"
                >
                  <span
                    class="icon-text padding-right-4 display-flex flex-align-center"
                  >
                    <svg
                      aria-label="Pencil icon indicating edit ability"
                      class="usa-icon height-3 width-3"
                      focusable="false"
                      height="1em"
                      role="img"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 0 0 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                      />
                    </svg>
                    <span
                      class="padding-left-05"
                      data-testid="edit-group-456"
                    >
                      Rename
                    </span>
                  </span>
                </button>
              </div>
            </td>
            <td
              class="roleDropdown"
            >
              <label
                class="usa-label usa-sr-only"
                data-testid="label"
                for="role-select-456"
              >
                User role
              </label>
              <select
                class="usa-select"
                data-testid="role-select-456"
                id="role-select-456"
              >
                <option
                  value="Super Admin"
                >
                  Super Admin
                </option>
                <option
                  selected=""
                  value="Admin"
                >
                  Admin
                </option>
                <option
                  value="Standard"
                >
                  Standard
                </option>
              </select>
            </td>
            <td>
              <button
                aria-description="Edit Order of the Phoenix members"
                class="usa-button usa-button--unstyled margin-right-2 text-no-underline"
                data-testid="button"
                type="button"
              >
                <span>
                  Order of the Phoenix
                </span>
              </button>
            </td>
          </tr>
          <tr
            id="123"
            tabindex="0"
          >
            <td>
              <div
                class="buttonHoverGroup"
              >
                <span>
                  Potter, Harry 
                </span>
                <button
                  class="usa-button usa-button--unstyled hoverBtn unstyled-button-container usa-button--unstyled text-bold text-no-underline"
                  data-testid="button"
                  type="button"
                >
                  <span
                    class="icon-text padding-right-4 display-flex flex-align-center"
                  >
                    <svg
                      aria-label="Pencil icon indicating edit ability"
                      class="usa-icon height-3 width-3"
                      focusable="false"
                      height="1em"
                      role="img"
                      viewBox="0 0 24 24"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04a.996.996 0 0 0 0-1.41l-2.34-2.34a.996.996 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                      />
                    </svg>
                    <span
                      class="padding-left-05"
                      data-testid="edit-group-123"
                    >
                      Rename
                    </span>
                  </span>
                </button>
              </div>
            </td>
            <td
              class="roleDropdown"
            >
              <label
                class="usa-label usa-sr-only"
                data-testid="label"
                for="role-select-123"
              >
                User role
              </label>
              <select
                class="usa-select"
                data-testid="role-select-123"
                id="role-select-123"
              >
                <option
                  selected=""
                  value="Super Admin"
                >
                  Super Admin
                </option>
                <option
                  value="Admin"
                >
                  Admin
                </option>
                <option
                  value="Standard"
                >
                  Standard
                </option>
              </select>
            </td>
            <td>
              --
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</body>
`;
